include/vinecopulib/misc/implementation/tools_eigen.ipp

```diff
@@ -75,13 +75,23 @@ check_if_in_unit_cube(const Eigen::MatrixXd& u)
 //! swap the columns of a two-column matrix
 //! @param u the matrix.
 //! @return a new matrix v with `v.col(0) = u.col(1)`, `v.col(1) = u.col(0)`.
-inline Eigen::Matrix<double, Eigen::Dynamic, 2>
-swap_cols(Eigen::Matrix<double, Eigen::Dynamic, 2> u)
+inline Eigen::MatrixXd
+swap_cols(Eigen::MatrixXd u)
 {
   u.col(0).swap(u.col(1));
   return u;
 }
 
+inline Eigen::VectorXd
+unique(const Eigen::VectorXd& x)
+{
+  std::vector<double> v(x.data(), x.data() + x.size());
+  std::sort(v.begin(), v.end()); // 1 1 2 2 3 3 3 4 4 5 5 6 7
```

Comment:
Why do you need to sort and then erase the elements from `v`? It seems like a lot of useless computations no? Or you need them sorted on the other side?

The reason that I ask is because using a hash-map would reduce to o(n) average time: https://www.geeksforgeeks.org/print-distinct-elements-given-integer-array/

Reply:
This is what I was thinking about:

```
inline Eigen::VectorXd
unique(const Eigen::VectorXd& x)
{
  // Creates an empty hashset
  std::unordered_set<double> s;
  
  size_t n = x.size();
  double tmp;
  // Traverse the input vector
  for (size_t i = 0; i < n; i++) {
    tmp = x(i);
    // If not present, then put it in the hashtable
    if (s.find(tmp) == s.end()) {
      s.insert(tmp);
    }
  }
  std::vector<double> v(s.begin(), s.end());
  return Eigen::Map<Eigen::VectorXd>(&v[0], v.size());
}
```